crossplane/builder.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+import os
+
+from .lexer import lex
+from .analyzer import analyze, enter_block_ctx
+from .errors import NgxParserDirectiveError
+from .compat import PY2, json
+
+DELIMITERS = ('{', '}', ';')
+
+
+def _escape(string):
+    prev, char = '', ''
+    for char in string:
+        if prev == '\\' or prev + char == '${':
+            prev += char
+            yield prev
+            continue
+        if prev == '$':
+            yield prev
+        if char not in ('\\', '$'):
+            yield char
+        prev = char
+    if char in ('\\', '$'):
+        yield char
+
+
+def _needs_quotes(string):
+    if string == '':
+        return True
+    elif string in DELIMITERS:
+        return False
+
+    # lexer should throw an error when variable expansion syntax
+    # is messed up, but just wrap it in quotes for now I guess
+    chars = _escape(string)
+
+    # arguments can't start with variable expansion syntax
+    char = next(chars)
+    if char.isspace() or char in ('{', ';', '"', "'", '${'):
+        return True
+
+    expanding = False
+    for char in chars:
+        if char.isspace() or char in ('{', ';', '"', "'"):
+            return True
+        elif char == ('${' if expanding else '}'):
+            return True
+        elif char == ('}' if expanding else '${'):
+            expanding = not expanding
+
+    return char in ('\\', '$') or expanding
+
+
+def _enquote(arg):
+    arg = str(arg.encode('utf-8') if PY2 else arg)
+    if _needs_quotes(arg):
+        arg = repr(arg.decode('string_escape') if PY2 else arg)
+    return arg
+
+
+def build(payload, indent=4, tabs=False):
+    padding = '\t' if tabs else ' ' * indent
+
+    def _build_lines(objs, depth):
```

Comment:
This is a really clever use of a generator.

Reply:
Thanks :)